### PR TITLE
Declare a new lint to properly deny warnings in rustdoc

### LINKED ISSUE #79792: rustdoc -D warnings doesn't fail for all warnings
#### I tried this code:

```bash
/// See [`foo:Foo`].
/// ```rust
fn main() {}

```

#### I expected to see this happen:
The env variable `-Dwarnings` is supposed to make rustdoc treat warnings as errors. For `export RUSTDOCFLAGS=""` the run below outputs only warnings, no errors. For `RUSTDOCFLAGS="-Dwarnings"` I would have expected both warnings to now be errors.

```
$ export RUSTDOCFLAGS="-Dwarnings"
$ cargo doc
 Documenting rustdoc-bug v0.1.0 (/tmp/bug/rustdoc-bug)
error: unresolved link to `foo:Foo`
 --> src/main.rs:1:10
  |
1 | /// See [`foo:Foo`].
  |          ^^^^^^^^^ no item named `foo:Foo` in scope
  |
  = note: `-D broken-intra-doc-links` implied by `-D warnings`
  = help: to escape `[` and `]` characters, add '\' before them like `\[` or `\]`

error: Rust code block is empty
 --> src/main.rs:2:5
  |
2 | /// ```rust
  |     ^^^^^^^

error: aborting due to previous error; 1 warning emitted```
```

#### Instead, this happened:
Only one of the warnings is switched to an error.

```
$ export RUSTDOCFLAGS="-Dwarnings"
$ cargo doc
 Documenting rustdoc-bug v0.1.0 (/tmp/bug/rustdoc-bug)
error: unresolved link to `foo:Foo`
 --> src/main.rs:1:10
  |
1 | /// See [`foo:Foo`].
  |          ^^^^^^^^^ no item named `foo:Foo` in scope
  |
  = note: `-D broken-intra-doc-links` implied by `-D warnings`
  = help: to escape `[` and `]` characters, add '\' before them like `\[` or `\]`

warning: Rust code block is empty
 --> src/main.rs:2:5
  |
2 | /// ```rust
  |     ^^^^^^^

error: aborting due to previous error; 1 warning emitted```
```

This is also problematic because for warnings the exit code is still `0`, so a CI setup might not detect the failure:

```
$ cat src/main.rs 
/// ```rust
fn main() {}
$ RUSTDOCFLAGS="-Dwarnings" cargo doc
warning: Rust code block is empty
 --> src/main.rs:1:5
  |
1 | /// ```rust
  |     ^^^^^^^

warning: 1 warning emitted

    Finished dev [unoptimized + debuginfo] target(s) in 0.00s
$ echo $?
0
```

### Meta

`rustc --version --verbose`:
```
rustc 1.50.0-nightly (1c389ffef 2020-11-24)
```

### PR DESCRIPTION
**Pull request by poliorcetics**
This declares a new lint: `INVALID_RUST_CODEBLOCK` that is used by `rustdoc` to properly follow `-D warnings` instead of unconditionally emitting a warning.

## Todo List

- [x] Declare lint.
- [x] Document lint (file: `src/doc/rustdoc/src/lints.md`).
- [x] Use lint in `rustdoc` (file: `src/librustdoc/passes/check_code_block_syntax.rs`, maybe others).
- [x] Write tests.
  - [x] Note: one for the behaviour of the new lint when the error is in a dependency, not the crate being tested (https://github.com/rust-lang/rust/pull/79816#discussion_r541471520)
- [x] Refactor things.

- [x] Wait on #80527.

---

Fix #79792.

@rustbot label T-rustdoc A-lint

### CODE REVIEW

File: compiler/rustc_lint_defs/src/builtin.rs [hunk 1]
```diff
@@ -1847,6 +1847,18 @@ declare_lint! {
     "codeblock attribute looks a lot like a known one"
 }
 
+declare_lint! {
+    /// The `invalid_rust_codeblock` lint detects Rust code blocks in
+    /// documentation examples that are invalid (e.g.: empty, not parsable as
```
Comment (jyn514) [inline compiler/rustc_lint_defs/src/builtin.rs:1852]: This is a little hard to parse for me.
```suggestion
    /// documentation examples that are invalid (e.g. empty, not parseable as
```

File: src/doc/rustdoc/src/lints.md [hunk 2]
```diff
@@ -285,6 +285,58 @@ warning: unclosed HTML tag `h1`
 
 warning: 2 warnings emitted
 ```
+                
+## invalid_rust_codeblock
+
+This lint **warns by default**. It detects Rust code blocks in documentation
+examples that are invalid (e.g.: empty, not parsable as Rust). For example:
```
Comment (jyn514) [inline src/doc/rustdoc/src/lints.md:292]:
```suggestion
examples that are invalid (e.g. empty, not parseable as Rust). For example:
```

File: src/doc/rustdoc/src/lints.md [hunk 3]
```diff
@@ -285,6 +285,58 @@ warning: unclosed HTML tag `h1`
...
+
+```rust
+/// Empty code block, with and without Rust:
+///
+/// ```
+/// ```
+///
+/// ```rust
+/// ```
+///
```
Comment (jyn514) [inline src/doc/rustdoc/src/lints.md:302]: I think this example might be a little much - it's pretty clear what the error would be and I think it's better to make the docs concise. I could be convinced otherwise though.

File: src/librustdoc/passes/check_code_block_syntax.rs [hunk 4]
```diff
@@ -57,40 +58,56 @@ impl<'a, 'tcx> SyntaxChecker<'a, 'tcx> {
...
-                        sp,
-                        "mark blocks that do not contain Rust code as text",
-                        String::from("```text"),
-                        Applicability::MachineApplicable,
-                    );
-                }
-
-                diag
+                // NOTE(poliorcetics): is 'item' always local here ? Or will this panic one day ?
+                let hir_id = self.cx.tcx.hir().local_def_id_to_hir_id(item.def_id.expect_local());
```
Comment (jyn514) [inline src/librustdoc/passes/check_code_block_syntax.rs:62]: It is *not* always local (e.g. if this is a re-export), I think we found why it wasn't a lint before. Fortunately, we don't need to check the syntax of other crates, so you can just return without doing anything if it's non-local.

See https://github.com/rust-lang/rust/issues/77230 for more discussion; it would also be good to add a test, since if the behavior changes it should be intentional.

**Review by jyn514 — COMMENTED** [review]
This looks about like what I imagined, thanks :)

File: src/doc/rustdoc/src/lints.md (hunk 3 above)
Comment (poliorcetics) [reply src/doc/rustdoc/src/lints.md:302]: This file being the reference for rustdoc-lints, I would prefer it to be as complete as possible, but that's not something critical IMO, just a nice thing

Comment (jyn514) [reply src/doc/rustdoc/src/lints.md:302]: It's the reference for which lints *exist*, but not for all warnings emitted by the lints. For example intra-doc links only have a couple of the warnings emitted even though there are many more forms of the error: https://doc.rust-lang.org/nightly/rustdoc/lints.html#broken_intra_doc_links

I don't feel super strongly though, up to you if you want to change it.

File: src/librustdoc/passes/check_code_block_syntax.rs [hunk 5]
```diff
@@ -47,51 +49,78 @@ impl<'a, 'tcx> SyntaxChecker<'a, 'tcx> {
...
+        // The span and whether it is precise or not.
+        let (sp, precise_span) = match super::source_span_for_markdown_range(
+            self.cx,
+            &dox,
+            &code_block.range,
+            &item.attrs,
+        ) {
+            Some(sp) => (sp, true),
+            None => {
+                (super::span_of_attrs(&item.attrs).unwrap_or_else(|| item.source.span()), false)
```
Comment (jyn514) [inline src/librustdoc/passes/check_code_block_syntax.rs:76]: FYI `span` is trivial, I think `or_else` is a little more verbose than it needs to be. https://doc.rust-lang.org/nightly/nightly-rustc/src/rustdoc/clean/types.rs.html#1629-1631
```suggestion
                (super::span_of_attrs(&item.attrs).unwrap_or(item.source.span()), false)
```

**Review by jyn514 — APPROVED** [review]
Thanks, LGTM. Can you squash the commits before merging?

**Comment by jyn514** [comment]
Also you need to update `src/test/rustdoc-ui`:

```
Some tests failed in compiletest suite=rustdoc-ui mode=ui host=x86_64-unknown-linux-gnu target=x86_64-unknown-linux-gnu
failures:

---- [ui] rustdoc-ui/invalid-syntax.rs stdout ----
diff of stderr:

7	LL | | /// ```
8	   | |_______^
9	   |
+	   = note: `#[warn(invalid_rust_codeblock)]` on by default
10	   = note: error from rustc: unknown start of token: \
11	   = note: error from rustc: unknown start of token: \
12	   = note: error from rustc: unknown start of token: \
```

**Comment by poliorcetics** [comment]
Added the changes you proposed and squashed the whole thing

File: src/doc/rustdoc/src/lints.md (hunk 3 above)
Comment (poliorcetics) [reply src/doc/rustdoc/src/lints.md:299]: I left one instance of an empty code block and one for an invalid code block

**Comment by jyn514** [comment]
@bors r+

Looks great, thanks!

**Comment by jyn514** [comment]
Now you get to fix the compiler docs :laughing: I guess they should be fixed to use `text` instead?

**Comment by poliorcetics** [comment]
Should be simple enough most of the time, I won't have time to work on it before Friday afternoon at the soonest though

**Comment by poliorcetics** [comment]
The failing example is:

```rust
    /// ```ignore (incomplete snippet)
    /// type Foo = impl Baz;
    /// fn bar() -> Foo {
    /// //          ^^^ This is the span we are looking for!
    /// ```
```

It's tricky because we still want Rust-highlighting but no compilation at all, is this feasible ?

**Comment by jyn514** [comment]
I'm surprised it's highlighted currently: https://doc.rust-lang.org/nightly/nightly-rustc/rustc_trait_selection/opaque_types/struct.OpaqueTypeDecl.html#structfield.definition_span - maybe rustc_lexer is a streaming parser?

I think the easiest fix is just to add a closing delimiter. You can hide it if you like, so it doesn't show up in the docs: `/// # }`

**Comment by poliorcetics** [comment]
I didn't find a way to run `check` for all doc tests, is there a way to do it using `x.py` ?

**Comment by poliorcetics** [comment]
I think there is no other test that is concerned by this change, this can be rolled up and merged.
I ran `./x.py test -i --doc --stage 1` to be sure.

**Comment by jyn514** [comment]
@bors r+

@poliorcetics 'rollup' is for PRs that should *always* be rolled up, which I don't think is the case here.

File: compiler/rustc_trait_selection/src/opaque_types.rs [hunk 6]
```diff
@@ -46,6 +46,7 @@ pub struct OpaqueTypeDecl<'tcx> {
     /// type Foo = impl Baz;
     /// fn bar() -> Foo {
     /// //          ^^^ This is the span we are looking for!
+    /// # }
```
Comment (GuillaumeGomez) [inline compiler/rustc_trait_selection/src/opaque_types.rs:49]: Why this change btw?

Comment (poliorcetics) [reply compiler/rustc_trait_selection/src/opaque_types.rs:49]: Without it the build fails because of -Dwarnings, see https://github.com/rust-lang/rust/pull/79816#issuecomment-748532010

Comment (GuillaumeGomez) [reply compiler/rustc_trait_selection/src/opaque_types.rs:49]: Then use `}` instead of `# }`, there is not much point into hiding this paren.

Comment (poliorcetics) [reply compiler/rustc_trait_selection/src/opaque_types.rs]: Done :)

**Comment by Dylan-DPC** [comment]
failed in [rollup log](https://github.com/rust-lang/rust/pull/79816)

**Comment by poliorcetics** [comment]
I'm officially lost in the hell that are git submodules

**Comment by poliorcetics** [comment]
Closing for #80456.

**Comment by poliorcetics** [comment]
As asked by @jyn514 I'll reopen here and close the new PR for easier review.

**Comment by jyn514** [comment]
r=me on the implementation, but I'd prefer to wait to hear about whether rustdoc should be adding its own lints without talking with T-lang (https://github.com/rust-lang/rust/pull/80456#issuecomment-751882904, https://rust-lang.zulipchat.com/#narrow/stream/131828-t-compiler/topic/What.20is.20the.20policy.20for.20new.20lints.3F).

**Comment by crlf0710** [comment]
Seems this is blocked on #80527. I'll mark this so.

**Comment by jyn514** [comment]
https://github.com/rust-lang/rust/pull/80527 was merged, so this is no longer blocked :)

@rustbot label: -S-blocked +S-waiting-on-author

File: compiler/rustc_lint_defs/src/builtin.rs [hunk 7]
```diff
@@ -2933,6 +3032,15 @@ declare_lint_pass! {
...
         IRREFUTABLE_LET_PATTERNS,
+        BROKEN_INTRA_DOC_LINKS,
+        PRIVATE_INTRA_DOC_LINKS,
+        INVALID_CODEBLOCK_ATTRIBUTES,
+        INVALID_RUST_CODEBLOCK,
+        MISSING_CRATE_LEVEL_DOCS,
+        MISSING_DOC_CODE_EXAMPLES,
+        INVALID_HTML_TAGS,
+        PRIVATE_DOC_TESTS,
+        NON_AUTOLINKS,
```
Comment (jyn514) [inline compiler/rustc_lint_defs/src/builtin.rs:3043]: This looks like a bad rebase, these should not be part of rustc proper.

File: compiler/rustc_lint_defs/src/builtin.rs [hunk 8]
```diff
@@ -1875,6 +1875,105 @@ declare_lint! {
...
+   /// The `invalid_rust_codeblock` lint detects Rust code blocks in
+   /// documentation examples that are invalid (e.g. empty, not parsable as
+   /// Rust code). This is a `rustdoc` only lint, see the documentation in the
+   /// [rustdoc book].
+   ///
+   /// [rustdoc book]: ../../../rustdoc/lints.html#invalid_rust_codeblock
+   pub INVALID_RUST_CODEBLOCK,
+   Warn,
+   "codeblock could not be parsed as valid Rust or is empty"
+}
```
Comment (jyn514) [inline compiler/rustc_lint_defs/src/builtin.rs:1921]: This should be declared in librustdoc/lint.rs, not here.

**Comment by jyn514** [comment]
This is still not correct. Any new lints should be added to `librustdoc/lint.rs`. `rustc_lint_defs` shouldn't be modified at all, that's for builtin lints, not tool lints.

File: src/test/rustdoc-ui/invalid-syntax.stderr [hunk 9]
```diff
@@ -7,6 +7,7 @@ LL | | /// \__________pkt->size___________/          \_result->size_/ \__pkt->si
 LL | | /// ```
    | |_______^
    |
+   = note: `#[warn(invalid_rust_codeblock)]` on by default
```
Comment (jyn514) [inline src/test/rustdoc-ui/invalid-syntax.stderr:10]: Could you add some tests for when `ignore` is present (and for all the rest of the branching you have for the diagnostics)?

File: src/librustdoc/passes/check_code_block_syntax.rs (hunk 4 above)
Comment (jyn514) [reply src/librustdoc/passes/check_code_block_syntax.rs]: Could you add a test for the behavior?

**Comment by JohnCSimon** [comment]
@poliorcetics - Ping from triage: can you please resolve the merge conflict and set it to `S-waiting-on-review` ?

**Comment by jyn514** [comment]
Hi @poliorcetics, any updates on this? It's really close to being done :/

**Comment by jyn514** [comment]
I'm going to close this - feel free to re-open if you have time to work on this again.
